system_test/aest_sync_SUITE.erl

```diff
@@ -0,0 +1,190 @@
+
+%=== EXPORTS ===================================================================
+
+% Common Test exports
+-export([all/0]).
+-export([init_per_testcase/2]).
+-export([end_per_testcase/2]).
+
+% Test cases
+-export([new_node_joins_network/1,
+         docker_keeps_data/1,
+         crash_and_continue_sync/1]).
+
+-import(aest_nodes, [
+    setup_nodes/2,
+    start_node/2,
+    stop_node/2, stop_node/3,
+    kill_node/2,
+    http_get/5,
+    request/4,
+    wait_for_height/4,
+    assert_synchronized/2
+]).
+
+%=== INCLUDES ==================================================================
+
+-include_lib("eunit/include/eunit.hrl").
+
+%=== MACROS ====================================================================
+
+% -define(MINING_TIMEOUT, 4*10000).
+-define(MINING_TIMEOUT, 1000).
+
+-define(NODE1, #{
+    name    => node1,
+    peers   => [node2],
+    backend => aest_docker,
+    source  => {pull, "aetrnty/epoch:v0.8.0"}
+}).
+
+-define(NODE2, #{
+    name    => node2,
+    peers   => [node1],
+    backend => aest_docker,
+    source  => {pull, "aetrnty/epoch:v0.8.0"}
+}).
+
+-define(NODE3, #{
+    name    => node3,
+    peers   => [node1],
+    backend => aest_docker,
+    source  => {pull, "aetrnty/epoch:local"}
+}).
+
+%=== COMMON TEST FUNCTIONS =====================================================
+
+all() -> [
+    new_node_joins_network
+    % , docker_keeps_data
+    , crash_and_continue_sync
+].
+
+init_per_testcase(_TC, Config) ->
+    %% Some parameters depend on the speed and capacity of the docker containers:
+    aest_nodes:ct_setup([ {blocks_per_second, 3},
+                          {node_startup_time, 5000}, %% Time it takes to get the node to respond to http
+                          {node_stop_time, 20000}    %% Time it takes to get the node to stop cleanly
+                          | Config]).
+
+end_per_testcase(_TC, Config) ->
+    aest_nodes:ct_cleanup(Config).
+
+%=== TEST CASES ================================================================
+
+%% A few tests that verify that our assumptions are right for docker timings and
+%% API.
+
+new_node_joins_network(Cfg) ->
+    Length = 10,
+    BlocksPerSecond = proplists:get_value(blocks_per_second, Cfg),
+    NodeStartupTime = proplists:get_value(node_startup_time, Cfg), 
+
+    setup_nodes([?NODE1, ?NODE2, ?NODE3], Cfg),
+    start_node(node1, Cfg),
+    start_node(node2, Cfg),
+    wait_for_height(Length, [node1, node2], Length * ?MINING_TIMEOUT, Cfg),
+    Top1 = request(node1, [v2, 'top'], #{}, Cfg),
+    ct:log("Node 1 top: ~p~n", [Top1]),
+    Height = request(node2, [v2, 'block-by-height'], #{height => Length}, Cfg),
+    ct:log("Node 2 at height ~p: ~p~n", [Length, Height]),
+    start_node(node3, Cfg),
+    wait_for_height(0, [node3], NodeStartupTime, Cfg),
+    ct:log("Node 3 ready to go"),
+
+    %% get them with N blocks a second
+    timer:sleep((Length div BlocksPerSecond) * 1000),
+
+    wait_for_height(Length, [node3], 1000, Cfg),
```

Comment:
I replaced `aetrnty/epoch:local` with `aeternity/epoch:dev` then I got:
```
=== Location: [{aest_nodes,assert_expiration,361},
              {aest_nodes,wait_for_height,367},
              {aest_sync_SUITE,new_node_joins_network,99},
```

Reply:
Maybe it is actually catching a real synchronization issue between `v0.8.0` and `dev`! It fails for me too, but if I set the 3 nodes to `dev` the test passes.

Reply:
I see in logs:
```
2018-03-26 16:04:19.825 [debug] <0.1111.0>@aeu_requests:ping:73 Got an error return: Reason = <<"Different genesis blocks">>
2018-03-26 16:04:19.825 [debug] <0.1111.0>@aec_sync:ping_peer:455 ping result (<<"http://node1_61A8E9F439C197A4:3013">>): {error,protocol_violation}
```
Maybe test is meant not to run v0.8.0 with master/`dev` but rather a previous version e.g. v0.8.0 itself? Or all :dev?

Reply:
This test is meant to ensure that the last version is able to sync with a cluster of older versions. If it does not pass, either we broke the compatibility and we know it (then we need to update the images) or we broke the protocol (or the test is bad to begin with).

Reply:
> Maybe it is actually catching a real synchronization issue between v0.8.0 and dev! It fails for me too, but if I set the 3 nodes to dev the test passes.

You are right. Recent versions are not sync compatible so this shall - at least at the moment - have all the same version.

----

Chain consensus protocol in release 0.8.0:
https://github.com/aeternity/epoch/blob/v0.8.0/apps/aecore/include/blocks.hrl#L3-L5

Chain consensus protocol in release 0.9.0:
https://github.com/aeternity/epoch/blob/v0.9.0/apps/aecore/include/blocks.hrl#L3-L5

Chain consensus protocol in current master:
https://github.com/aeternity/epoch/blob/e65c7e4d838db29b52f371ab97c96eaae84bf717/apps/aecore/include/blocks.hrl#L3-L5